File: backend/app/schemas/occurrences.py
"""Occurrence schemas — Phase 4 Core Tracking Engine."""

from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class OccurrenceBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class Occurrence(OccurrenceBase):
    id: UUID
    habit_id: UUID
    habit_name: str
    occurrence_date: date
    status: Literal["pending", "completed", "skipped", "missed"]
    completed_at: datetime | None = None
    reminder_time: str | None = None


class Progress(BaseModel):
    completed: int
    total: int
    percent: int


class TodayResponse(BaseModel):
    date: date
    occurrences: list[Occurrence]
    progress: Progress


class OccurrenceListResponse(BaseModel):
    occurrences: list[Occurrence]


class OccurrenceUpdate(BaseModel):
    status: Literal["completed", "skipped", "pending"]
    model_config = ConfigDict(extra="forbid")

File: mobile/src/api/occurrences.ts
import { apiClient } from './client';

export type OccurrenceStatus = 'pending' | 'completed' | 'skipped' | 'missed';

export interface Occurrence {
  id: string;
  habit_id: string;
  habit_name: string;
  occurrence_date: string;
  status: OccurrenceStatus;
  completed_at: string | null;
  reminder_time: string | null;
}

export interface Progress {
  completed: number;
  total: number;
  percent: number;
}

export interface TodayResponse {
  date: string;
  occurrences: Occurrence[];
  progress: Progress;
}

export const fetchToday = async (): Promise<TodayResponse> => {
  const response = await apiClient.get<TodayResponse>('/api/v1/today');
  return response.data;
};

export const updateOccurrence = async (
  id: string,
  status: OccurrenceStatus,
  idempotencyKey: string
): Promise<Occurrence> => {
  const response = await apiClient.put<Occurrence>(
    `/api/v1/occurrences/${id}`,
    { status },
    {
      headers: {
        'Idempotency-Key': idempotencyKey,
      },
    }
  );
  return response.data;
};
